// cosmjs-types/cosmos/base/store/v1beta1/listening.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockMetadata_DeliverTx = exports.BlockMetadata = exports.StoreKVPair = exports.protobufPackage = void 0;
/* eslint-disable */
const types_1 = require("../../../../tendermint/abci/types");
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "cosmos.base.store.v1beta1";
function createBaseStoreKVPair() {
    return {
        storeKey: "",
        delete: false,
        key: new Uint8Array(),
        value: new Uint8Array(),
    };
}
exports.StoreKVPair = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.storeKey !== "") {
            writer.uint32(10).string(message.storeKey);
        }
        if (message.delete === true) {
            writer.uint32(16).bool(message.delete);
        }
        if (message.key.length !== 0) {
            writer.uint32(26).bytes(message.key);
        }
        if (message.value.length !== 0) {
            writer.uint32(34).bytes(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStoreKVPair();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.storeKey = reader.string();
                    break;
                case 2:
                    message.delete = reader.bool();
                    break;
                case 3:
                    message.key = reader.bytes();
                    break;
                case 4:
                    message.value = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            storeKey: (0, helpers_1.isSet)(object.storeKey) ? String(object.storeKey) : "",
            delete: (0, helpers_1.isSet)(object.delete) ? Boolean(object.delete) : false,
            key: (0, helpers_1.isSet)(object.key) ? (0, helpers_1.bytesFromBase64)(object.key) : new Uint8Array(),
            value: (0, helpers_1.isSet)(object.value) ? (0, helpers_1.bytesFromBase64)(object.value) : new Uint8Array(),
        };
    },
    toJSON(message) {
        const obj = {};
        message.storeKey !== undefined && (obj.storeKey = message.storeKey);
        message.delete !== undefined && (obj.delete = message.delete);
        message.key !== undefined &&
            (obj.key = (0, helpers_1.base64FromBytes)(message.key !== undefined ? message.key : new Uint8Array()));
        message.value !== undefined &&
            (obj.value = (0, helpers_1.base64FromBytes)(message.value !== undefined ? message.value : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = createBaseStoreKVPair();
        message.storeKey = object.storeKey ?? "";
        message.delete = object.delete ?? false;
        message.key = object.key ?? new Uint8Array();
        message.value = object.value ?? new Uint8Array();
        return message;
    },
};
function createBaseBlockMetadata() {
    return {
        requestBeginBlock: undefined,
        responseBeginBlock: undefined,
        deliverTxs: [],
        requestEndBlock: undefined,
        responseEndBlock: undefined,
        responseCommit: undefined,
    };
}
exports.BlockMetadata = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.requestBeginBlock !== undefined) {
            types_1.RequestBeginBlock.encode(message.requestBeginBlock, writer.uint32(10).fork()).ldelim();
        }
        if (message.responseBeginBlock !== undefined) {
            types_1.ResponseBeginBlock.encode(message.responseBeginBlock, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.deliverTxs) {
            exports.BlockMetadata_DeliverTx.encode(v, writer.uint32(26).fork()).ldelim();
        }
        if (message.requestEndBlock !== undefined) {
            types_1.RequestEndBlock.encode(message.requestEndBlock, writer.uint32(34).fork()).ldelim();
        }
        if (message.responseEndBlock !== undefined) {
            types_1.ResponseEndBlock.encode(message.responseEndBlock, writer.uint32(42).fork()).ldelim();
        }
        if (message.responseCommit !== undefined) {
            types_1.ResponseCommit.encode(message.responseCommit, writer.uint32(50).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBlockMetadata();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.requestBeginBlock = types_1.RequestBeginBlock.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.responseBeginBlock = types_1.ResponseBeginBlock.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.deliverTxs.push(exports.BlockMetadata_DeliverTx.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.requestEndBlock = types_1.RequestEndBlock.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.responseEndBlock = types_1.ResponseEndBlock.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.responseCommit = types_1.ResponseCommit.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            requestBeginBlock: (0, helpers_1.isSet)(object.requestBeginBlock)
                ? types_1.RequestBeginBlock.fromJSON(object.requestBeginBlock)
                : undefined,
            responseBeginBlock: (0, helpers_1.isSet)(object.responseBeginBlock)
                ? types_1.ResponseBeginBlock.fromJSON(object.responseBeginBlock)
                : undefined,
            deliverTxs: Array.isArray(object?.deliverTxs)
                ? object.deliverTxs.map((e) => exports.BlockMetadata_DeliverTx.fromJSON(e))
                : [],
            requestEndBlock: (0, helpers_1.isSet)(object.requestEndBlock)
                ? types_1.RequestEndBlock.fromJSON(object.requestEndBlock)
                : undefined,
            responseEndBlock: (0, helpers_1.isSet)(object.responseEndBlock)
                ? types_1.ResponseEndBlock.fromJSON(object.responseEndBlock)
                : undefined,
            responseCommit: (0, helpers_1.isSet)(object.responseCommit)
                ? types_1.ResponseCommit.fromJSON(object.responseCommit)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.requestBeginBlock !== undefined &&
            (obj.requestBeginBlock = message.requestBeginBlock
                ? types_1.RequestBeginBlock.toJSON(message.requestBeginBlock)
                : undefined);
        message.responseBeginBlock !== undefined &&
            (obj.responseBeginBlock = message.responseBeginBlock
                ? types_1.ResponseBeginBlock.toJSON(message.responseBeginBlock)
                : undefined);
        if (message.deliverTxs) {
            obj.deliverTxs = message.deliverTxs.map((e) => (e ? exports.BlockMetadata_DeliverTx.toJSON(e) : undefined));
        }
        else {
            obj.deliverTxs = [];
        }
        message.requestEndBlock !== undefined &&
            (obj.requestEndBlock = message.requestEndBlock
                ? types_1.RequestEndBlock.toJSON(message.requestEndBlock)
                : undefined);
        message.responseEndBlock !== undefined &&
            (obj.responseEndBlock = message.responseEndBlock
                ? types_1.ResponseEndBlock.toJSON(message.responseEndBlock)
                : undefined);
        message.responseCommit !== undefined &&
            (obj.responseCommit = message.responseCommit
                ? types_1.ResponseCommit.toJSON(message.responseCommit)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseBlockMetadata();
        message.requestBeginBlock =
            object.requestBeginBlock !== undefined && object.requestBeginBlock !== null
                ? types_1.RequestBeginBlock.fromPartial(object.requestBeginBlock)
                : undefined;
        message.responseBeginBlock =
            object.responseBeginBlock !== undefined && object.responseBeginBlock !== null
                ? types_1.ResponseBeginBlock.fromPartial(object.responseBeginBlock)
                : undefined;
        message.deliverTxs = object.deliverTxs?.map((e) => exports.BlockMetadata_DeliverTx.fromPartial(e)) || [];
        message.requestEndBlock =
            object.requestEndBlock !== undefined && object.requestEndBlock !== null
                ? types_1.RequestEndBlock.fromPartial(object.requestEndBlock)
                : undefined;
        message.responseEndBlock =
            object.responseEndBlock !== undefined && object.responseEndBlock !== null
                ? types_1.ResponseEndBlock.fromPartial(object.responseEndBlock)
                : undefined;
        message.responseCommit =
            object.responseCommit !== undefined && object.responseCommit !== null
                ? types_1.ResponseCommit.fromPartial(object.responseCommit)
                : undefined;
        return message;
    },
};
function createBaseBlockMetadata_DeliverTx() {
    return {
        request: undefined,
        response: undefined,
    };
}
exports.BlockMetadata_DeliverTx = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.request !== undefined) {
            types_1.RequestDeliverTx.encode(message.request, writer.uint32(10).fork()).ldelim();
        }
        if (message.response !== undefined) {
            types_1.ResponseDeliverTx.encode(message.response, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBlockMetadata_DeliverTx();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.request = types_1.RequestDeliverTx.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.response = types_1.ResponseDeliverTx.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            request: (0, helpers_1.isSet)(object.request) ? types_1.RequestDeliverTx.fromJSON(object.request) : undefined,
            response: (0, helpers_1.isSet)(object.response) ? types_1.ResponseDeliverTx.fromJSON(object.response) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.request !== undefined &&
            (obj.request = message.request ? types_1.RequestDeliverTx.toJSON(message.request) : undefined);
        message.response !== undefined &&
            (obj.response = message.response ? types_1.ResponseDeliverTx.toJSON(message.response) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseBlockMetadata_DeliverTx();
        message.request =
            object.request !== undefined && object.request !== null
                ? types_1.RequestDeliverTx.fromPartial(object.request)
                : undefined;
        message.response =
            object.response !== undefined && object.response !== null
                ? types_1.ResponseDeliverTx.fromPartial(object.response)
                : undefined;
        return message;
    },
};
